TailLight/SetTaillightBlack.cpp

```diff
@@ -0,0 +1,269 @@
+        WDF_OBJECT_ATTRIBUTES_INIT(&workItemAttributes);
+        WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&workItemAttributes,
+            SET_BLACK_WORKITEM_CONTEXT);
+        workItemAttributes.ParentObject = device;
+
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+
+        // It's possible to get called twice. Please refer to 
+        // https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-ioregisterplugplaynotification,
+        // under "Remarks"
+        if ((!pDeviceContext) || pDeviceContext->fSetBlackSuccess) {
+            return STATUS_SUCCESS;
+        }
+
+        WDF_WORKITEM_CONFIG_INIT(&workItemConfig, SetBlackWorkItem);
+
+        status = WdfWorkItemCreate(&workItemConfig,
+            &workItemAttributes,
+            &hWorkItem);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrint(("TailLight: Workitem creation failure 0x%x\n",
+                status));
+            return status; // Maybe better luck next time.
+        }
+
+        auto workItemContext = WdfObjectGet_SET_BLACK_WORKITEM_CONTEXT(hWorkItem);
+        workItemContext->symLink = symLink;
+    }
+
+    WdfWorkItemEnqueue(hWorkItem);
+
+    return status;
+}
+
+static NTSTATUS TryToOpenIoTarget(WDFIOTARGET target, 
+    CONST UNICODE_STRING symLink) {
+
+    PAGED_CODE();
+
+    WDF_IO_TARGET_OPEN_PARAMS   openParams = {};
+    WDF_IO_TARGET_OPEN_PARAMS_INIT_OPEN_BY_NAME(
+        &openParams,
+        &symLink,
+        STANDARD_RIGHTS_ALL);
+
+    openParams.ShareAccess = FILE_SHARE_WRITE | FILE_SHARE_READ;
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    // Ensure freed if fails.
+    status = WdfIoTargetOpen(target, &openParams);
+    if (!NT_SUCCESS(status)) {
+        KdPrint(("TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+    }
+
+    return status;
+}
+
+void SetBlackCompletionRoutine(
+    _In_ WDFREQUEST Request,
+    _In_ WDFIOTARGET Target,
+    _In_ PWDF_REQUEST_COMPLETION_PARAMS Params,
+    _In_ WDFCONTEXT Context)
+{
+    UNREFERENCED_PARAMETER(Target);
+    UNREFERENCED_PARAMETER(Params);
+    UNREFERENCED_PARAMETER(Context);
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    status = WdfRequestGetStatus(Request);
+    KdPrint(("TailLight: %s WdfRequestSend status: 0x%x\n", __func__, status));
+
+    // One-shot and top of stack, so delete.
+    WdfObjectDelete(Request);
+}
+
+VOID SetBlackWorkItem(
+    WDFWORKITEM workItem)
+    /*++
+
+    Routine Description:
+
+        Creates a WDF workitem to do the SetBlack() call after the driver
+        stack has initialized.
+
+    Arguments:
+
+        workItem - Handle to a pre-allocated WDF work item.
+    --*/
+{
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    WDFDEVICE device = static_cast<WDFDEVICE>(WdfWorkItemGetParentObject(workItem));
+    auto workItemContext = WdfObjectGet_SET_BLACK_WORKITEM_CONTEXT(workItem);
+
+    status = SetBlackAsync(device, workItemContext->symLink);
```

Comment:
Please add logging if the `SetBlackAsync` call fails.

I'm observing >10 failing  `SetBlackAsync` calls when attaching a mouse when testing this PR. Still, the tail-light is cleared, so the function "works" from a functionality point-of-view. Still, these errors will need to be fixed.

Reply:
It seems to me like `SetBlackAsync` might still fail silently without any logging if the `WdfRequestSend` call within is failing. Please ensure that all failure cases are logged properly, so that they do not go unnoticed.

Reply:
Thanks for addressing this comment.